// @langchain/openrouter/dist/utils/errors.js
import { LangChainError, ns } from "@langchain/core/errors";
//#region src/utils/errors.ts
const ns$1 = ns.sub("openrouter");
/**
* Base error for all OpenRouter API errors.
*
* Subclasses ({@link OpenRouterAuthError}, {@link OpenRouterRateLimitError})
* represent specific failure modes. Use the static {@link fromResponse}
* factory to let the library pick the right subclass from an HTTP response,
* or throw this class directly for generic API failures.
*
* Type-check errors with `OpenRouterError.isInstance(err)` rather than
* `instanceof` — this avoids cross-realm / duplicate-package pitfalls.
*/
var OpenRouterError = class OpenRouterError extends ns$1.brand(LangChainError) {
	name = "OpenRouterError";
	/** HTTP or API error code, if available. */
	code;
	/** Additional error metadata returned by the API, if available. */
	metadata;
	constructor(message, code, metadata) {
		super(message);
		this.code = code;
		this.metadata = metadata;
	}
	/**
	* Creates a typed error from an HTTP `Response`.
	*
	* Attempts to parse the body as JSON (OpenRouter's standard
	* `{ error: { message, code, metadata } }` shape). Falls back to the
	* raw HTTP status text when the body is missing or unparseable.
	*
	* Status-code mapping:
	* - 401 / 403 -> {@link OpenRouterAuthError}
	* - 429       -> {@link OpenRouterRateLimitError}
	* - anything else -> {@link OpenRouterError}
	*/
	static async fromResponse(response) {
		let body = {};
		try {
			body = await response.json();
		} catch {}
		const error = body?.error;
		const message = `${error?.message ?? `HTTP ${response.status}: ${response.statusText}`}${error?.metadata ? ` | metadata: ${JSON.stringify(error.metadata)}` : ""}`;
		const code = error?.code ?? response.status;
		if (response.status === 401 || response.status === 403) return new OpenRouterAuthError(message, code, error?.metadata);
		if (response.status === 429) return new OpenRouterRateLimitError(message, code, error?.metadata);
		return new OpenRouterError(message, code, error?.metadata);
	}
};
/**
* Thrown on authentication or authorization failures.
*
* Created automatically by {@link OpenRouterError.fromResponse} for
* HTTP 401/403 responses, and thrown directly by the constructor when
* no API key is provided.
*/
var OpenRouterAuthError = class extends ns$1.brand(OpenRouterError, "auth") {
	name = "OpenRouterAuthError";
	constructor(message, code, metadata) {
		super(message, code, metadata);
	}
};
/**
* Thrown when the API returns HTTP 429 (Too Many Requests).
*
* Callers can catch this specifically to implement back-off / retry
* logic without catching unrelated API errors.
*/
var OpenRouterRateLimitError = class extends ns$1.brand(OpenRouterError, "rate_limit") {
	name = "OpenRouterRateLimitError";
	constructor(message, code, metadata) {
		super(message, code, metadata);
	}
};
//#endregion
export { OpenRouterAuthError, OpenRouterError, OpenRouterRateLimitError };
